src/Service/ProducerEventProcessor.cpp

```diff
@@ -0,0 +1,371 @@
+  Callstack* callstack = full_callstack_sample->mutable_callstack();
+  auto [callstack_id, assigned] =
+      callstack_cache_.GetOrAssignId({callstack->pcs().begin(), callstack->pcs().end()});
+
+  if (assigned) {
+    ClientCaptureEvent interned_callstack_event;
+    interned_callstack_event.mutable_interned_callstack()->set_key(callstack_id);
+    *interned_callstack_event.mutable_interned_callstack()->mutable_intern() =
+        std::move(*callstack);
+    capture_event_buffer_->AddEvent(std::move(interned_callstack_event));
+  }
+
+  ClientCaptureEvent callstack_sample_event;
+  CallstackSample* callstack_sample = callstack_sample_event.mutable_callstack_sample();
+  callstack_sample->set_pid(full_callstack_sample->pid());
+  callstack_sample->set_tid(full_callstack_sample->tid());
+  callstack_sample->set_timestamp_ns(full_callstack_sample->timestamp_ns());
+  callstack_sample->set_callstack_id(callstack_id);
+  capture_event_buffer_->AddEvent(std::move(callstack_sample_event));
+}
+
+void ProducerEventProcessorImpl::ProcessInternedCallstack(uint64_t producer_id,
+                                                          InternedCallstack* interned_callstack) {
+  CHECK(!producer_interned_callstack_id_to_client_callstack_id_.contains(
+      {producer_id, interned_callstack->key()}));
+
+  std::vector<uint64_t> callstack_data{interned_callstack->intern().pcs().begin(),
+                                       interned_callstack->intern().pcs().end()};
+  auto [interned_callstack_id, assigned] = callstack_cache_.GetOrAssignId(callstack_data);
+
+  producer_interned_callstack_id_to_client_callstack_id_.insert_or_assign(
+      {producer_id, interned_callstack->key()}, interned_callstack_id);
+
+  if (!assigned) {
+    return;
+  }
+
+  // If this is first time we see it -> send it over with client_id
+  interned_callstack->set_key(interned_callstack_id);
+  ClientCaptureEvent event;
+  *event.mutable_interned_callstack() = std::move(*interned_callstack);
+  capture_event_buffer_->AddEvent(std::move(event));
+}
+
+void ProducerEventProcessorImpl::ProcessInternedCallstackSample(uint64_t producer_id,
+                                                                CallstackSample* callstack_sample) {
+  // translate producer id to client id
+  auto it = producer_interned_callstack_id_to_client_callstack_id_.find(
+      {producer_id, callstack_sample->callstack_id()});
+  CHECK(it != producer_interned_callstack_id_to_client_callstack_id_.end());
+  callstack_sample->set_callstack_id(it->second);
+
+  ClientCaptureEvent event;
+  *event.mutable_callstack_sample() = std::move(*callstack_sample);
+  capture_event_buffer_->AddEvent(std::move(event));
+}
+
+void ProducerEventProcessorImpl::ProcessInternedString(uint64_t producer_id,
+                                                       InternedString* interned_string) {
+  CHECK(!producer_interned_string_id_to_client_string_id_.contains(
+      {producer_id, interned_string->key()}));
```

Comment:
I'm skeptical about this `CHECK`. What if the producer, possibly a different process, makes a mistake and sends an intern twice? Should we really crash? The same for `ProcessInternedCallstack`.

Reply:
I was debating it. we can safely ignore it here, but in this case we won't know if producer starts sending duplicates. Just logging such event won't really have an effect, as a general rule - nobody reads logs. this is why I landed on the side of CHECKing this.